File: TwoPointers/CPP/di_string_match.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> diStringMatch(string s) {
        int i = 0, j = s.size();
        vector<int> ans;
        for (char c : s) {
            if (c == 'D') {
                ans.push_back(j);
                j--;
            } else {
                ans.push_back(i);
                i++;
            }
        }
        ans.push_back(i);
        return ans;
    }

    vector<int> diStringMatch(string S) {
        vector<int> res;
        for (int l = 0, h = S.size(), i = 0; i <= S.size(); ++i)
            res.push_back(i == S.size() || S[i] == 'I' ? l++ : h--);
    return res;
    }
};

File: TwoPointers/Python/di_string_match.py
class Solution:
    def diStringMatch(self, s: str) -> list[int]:
        i, j, ans = 0, len(s), []
        for char in s:
            if char == 'D':
                ans.append(j)
                j -= 1
            else:
                ans.append(i)
                i += 1
        ans.append(i)
        return ans